expression/constraint_propagation.go

```diff
@@ -164,3 +171,134 @@ func ruleColumnEQConst(ctx sessionctx.Context, i, j int, exprs *exprSet) {
 		}
 	}
 }
+
+// ruleColumnOPConst propagates the "column OP const" condition.
+func ruleColumnOPConst(ctx sessionctx.Context, i, j int, exprs *exprSet) {
+	cond := exprs.data[i]
+	f1, ok := cond.(*ScalarFunction)
+	if !ok {
+		return
+	}
+	if f1.FuncName.L != ast.GE && f1.FuncName.L != ast.GT &&
+		f1.FuncName.L != ast.LE && f1.FuncName.L != ast.LT {
+		return
+	}
+	OP1 := f1.FuncName.L
+
+	var col1 *Column
+	var con1 *Constant
+	col1, ok = f1.GetArgs()[0].(*Column)
+	if !ok {
+		return
+	}
+	con1, ok = f1.GetArgs()[1].(*Constant)
+	if !ok {
+		return
+	}
```

Comment:
We don't consider `const op col`?

Reply:
There are only two rules that really useful to partition pruning:

1. col < const
2. col >= const

Why?

```
create table ... partition by col 
p1 values less than const1
p2 values less than const2
...
```

Let's take a look at `p2`, the filter expressions are:

```
col >= const1 and col < const2
```

The final conditions would be:

```
col >= const1 and col < const2 and query
```

So the problem become that whether we can propagate:

```
col >= const1 and query => false
col < const2 and query => false
```

`ruleColumnOPConst` means use `col op const` as rule, perfectly cover the above two.

----------

You'll find the PR can't handle the `const op col`.
It doesn't matter, we can add another rule that transform
`const op col` to `col op const`

Then you'll find the rule order matters.

Not a serious problem, we can control the order by 
`newConstraintSolver(ruleTransformConstOPColumn, ruleColumnOPConst)`

I'll make pruning for  `select ... where to_days(c) > '2018-01-04'` work, and improve some cases later.